addon/services/stack-provider.js

```diff
@@ -0,0 +1,22 @@
+import Service from '@ember/service';
+import { action } from '@ember/object';
+
+export default class StackProviderService extends Service {
+  stack = [];
+
+  @action
+  hasOpenChild(dialog) {
```

Comment:
This is checking that the `dialog` argument is the currently-active dialog, yeah?

The name of this method is throwing me off a little -- it sort of suggests that it's checking if the `dialog` argument has a child that is open, rather than that the `dialog` argument is _itself_ open.

Reply:
Hmm I think that this method does exactly what you said: 

> it sort of suggests that it's checking if the `dialog` argument has a child that is open

if a dialog has an active (opened) child - which is the last element added to the stack, it will return false, otherwise true;

Reply:
Maybe turning it around to be something more like: `isCurrentlyOpenDialog` which checks that the provided dialog uid _is_ the last one in the stack? This might be less cognitive load.

Reply:
hmm this is the oposit logic, so it will change:
```js
if (this.dialogStackProvider.isCurrentlyOpenDialog(this)) {
   this.args.onClose();
}
```

But If 3 dialogs are open (nested dialogs) ? The name of this method does not make sense, right ? I don't fully get what is wrong with `hasOpenChild` 🤔

Reply:
My confusion with `hasOpenChild` is that is suggests that the `dialog` argument passed to it is not itself the top-most dialog. It reads to me like the method is checking that the `dialog` being passed in itself has some child dialog, which is open, but I don't think that's the intention.

It might be more clear to adopting the naming of typical Stack operations: `push`, `pop` and `peek`. Then if you wanted to check if a given dialog is the top-most one, you can do this instead:

```javascript
if (this.dialogStackProvider.peek() === this) {
  this.args.onClose()
}
```

The benefit here is that there's no longer ambiguity around what the intention of the code is; `peek` returns the top of a Stack, and we only want to operate if we're working with the `Dialog` at the top of the stack. The service then also conceptually makes to being just a stack without really needing any domain-specific behaviors, which further simplifies things.